File: ruby/quicksort.rb
def quicksort(arr)
  quick_helper(arr,0,arr.size-1)
  arr
end


def quick_helper(arr,s,e)
 return if s >= e
 pivot = s
 left = s+1
 right = e
  while left <= right
    swap(arr,left,right)  if arr[left] > arr[pivot] && arr[right] < arr[pivot]
    left+=1  if arr[left]<=arr[pivot]    
    right-=1 if arr[right]>=arr[pivot] 
  end
  swap(arr,pivot,right)#We swap the left most element with right
  is_small_rray = right-1-s > left+(right+1)
  quick_helper(arr,right+1,e)
  quick_helper(arr,s,right-1)
end

def swap(arr,l,r)
  arr[l],arr[r] = arr[r],arr[l]
end

 
p quicksort([8,12,14,5,2,9,5,6,3])

File: Js/quickSort.js
const quicksort = (arr) => {
  quickHelper(arr, 0, arr.length - 1)
  return arr
}

const quickHelper = (arr, start, end) => {
  if (start >= end) return;
  let pivot = start
  let left = start + 1
  let right = end
  while (start <= end) {
    if (arr[left] > arr[pivot] && arr[right] < arr[pivot]) {
      swap(arr, left, right)
    }
    if (arr[left] <= arr[pivot]) {
      left += 1
    }
    if (arr[right] >= arr[pivot]) {
      right -= 1
    }
  }
  swap(arr, pivot, right)

  quickHelper(arr, right + 1, end)

  quickHelper(arr, start, right - 1)
}

const swap = (arr, i, j) => {
  [arr[i], arr[j]] = [arr[j], arr[i]]
}


let s = quicksort([2, 3, 1, 9, 6, 2, 8])
console.log(s);